---
title: "Pushing latest sample summary to synapse"
author: "Camilo Posso"
date: "10/28/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(dplyr)
library(openxlsx)
source("../util/synapseUtil.R")
```



We just received the latest sample summary table, containing a lot of annotation
data, including allele ratios. 



```{r loading data}
path <- "../proteomics/data/BeatAML_complete_sample_summary/PNNL_clinical_summary_12_08_2021_updated_OS_4patients_02_28_2022.xlsx"
summary.table <- read.xlsx(path, sheet = 1)
write.table(summary.table, 
            "../proteomics/data/BeatAML_complete_sample_summary/PNNL_clinical_summary_12_08_2021_updated_OS_4patients_02_28_2022.txt")
path <- "../proteomics/data/BeatAML_complete_sample_summary/PNNL_clinical_summary_12_08_2021_updated_OS_4patients_02_28_2022.txt"

## Uploading to the "Genomics transcriptomics clinical" folder, under 
## the BeatAML Patient Data main folder.
parentID <- "syn24171152"

synapseStore(path, parentID)

```




